django/db/models/fields/__init__.py

```diff
@@ -678,6 +675,12 @@ def contribute_to_class(self, cls, name, virtual_only=False):
             cls._meta.add_field(self, virtual=True)
         else:
             cls._meta.add_field(self)
+        if self.column:
+            # Refuse to override class methods with the descriptor. This leads to a situation
+            # where if you have a classmethod and a field similarly named, then such fields
+            # can't be deferred (we don't have a check for this yet).
+            if not getattr(cls, self.attname, None):
+                setattr(cls, self.attname, DeferredAttribute(self.attname, cls))
```

Comment:
I think you're right about `__get__` useless checking of `instance.__dict__[self.field_name]`.

There's a lot of duplication between what `DeferredAttribute` does now and `cached_property`. I wonder if we could simply make `DeferredAttribute` a subclass of in this case.